src/Data/Singletons/Promote/Defun.hs

```diff
@@ -340,279 +386,496 @@ defunctionalize name m_fixity m_arg_tvbs' m_res_kind' = do
-         (d) [(y :: Proxy t)] (`k`, `t` and `a` were already bound)
+  type (*>@#@$$$) :: f a -> f b -> f b
+  type (*>@#@$$$) x y = (*>) x y
 
-         Therefore, we end up with:
+It turns out, however, that (*>@#@$$$) will not kind-check. Because (*>@#@$$$)
+has a standalone kind signature, it is kind-generalized *before* kind-checking
+the actual definition itself. Therefore, the full kind is:
 
-           data SymmetrySym1 (a :: Proxy (t :: k))
-             :: forall (y :: Proxy t).
-                            Proxy t
-                         ~> ((a :: Proxy (t :: k)) :~: (y :: Proxy (t :: k)))
-                         ~> Type
--}
+  type (*>@#@$$$) :: forall {k} (f :: k -> Type) (a :: k) (b :: k).
+                     f a -> f b -> f b
+  type (*>@#@$$$) x y = (*>) x y
 
--- This is a small function with large importance. When generating
--- defunctionalization data types, we often need to fill in the blank in the
--- sort of code exemplified below:
---
--- @
--- data FooSym2 a (b :: x) (c :: TyFun y z) where
---   FooSym2KindInference :: _
--- @
---
--- Where the kind of @a@ is not known. It's extremely tempting to just
--- copy-and-paste the type variable binders from the data type itself to the
--- constructor, like so:
---
--- @
--- data FooSym2 a (b :: x) (c :: TyFun y z) where
---   FooSym2KindInference :: forall a (b :: x) (c :: TyFun y z).
---                           SameKind (...) (...).
---                           FooSym2KindInference a b c
--- @
---
--- But this ends up being an untenable approach. Because @a@ lacks a kind
--- signature, @FooSym2@ does not have a complete, user-specified kind signature
--- (or CUSK), so GHC will fail to typecheck @FooSym2KindInference@.
---
--- Thankfully, there's a workaround—just don't give any of the constructor's
--- type variable binders any kinds:
---
--- @
--- data FooSym2 a (b :: x) (c :: TyFun y z) where
---   FooSym2KindInference :: forall a b c
---                           SameKind (...) (...).
---                           FooSym2KindInference a b c
--- @
---
--- GHC may be moody when it comes to CUSKs, but it's at least understanding
--- enough to typecheck this without issue. The 'dropTvbKind' function is
--- what removes the kinds used in the kind inference constructor.
-dropTvbKind :: DTyVarBndr -> DTyVarBndr
-dropTvbKind tvb@(DPlainTV {}) = tvb
-dropTvbKind (DKindedTV n _)   = DPlainTV n
+However, the kind of (*>) is
+`forall (f :: Type -> Type) (a :: Type) (b :: Type). f a -> f b -> f b`.
+This is not general enough for (*>@#@$$$), which expects kind-polymorphic `f`,
+`a`, and `b`, leading to a kind error. You might think that we could somehow
+infer this information, but note the quoted definition of Applicative (and
+PApplicative, as a consequence) omits the kinds of `f`, `a`, and `b` entirely.
+Unless we were to implement full-blown kind inference inside of Template
+Haskell (which is a tall order), the kind `f a -> f b -> f b` is about as good
+as we can get.
```

Comment:
Why doesn't this problem happen for unsaturated symbols? I don't see how saturation causes a difference here. Or maybe it does happen for unsaturated symbols, but no one notices...

Reply:
The difference is that unsaturated defunctionalization symbols are implemented as GADTs, and GADTs let you match on kind arguments. For instance, in the definition of `(*>@#@$)`:

```hs
type (*>@#@$) :: f a ~> f b ~> f b
data (*>@#@$) f where
  (*>@#@$###) :: forall f z. SameKind (Apply (*>@#@$) z) ((*>@#@$$) z) =>
                        (*>@#@$) f
```

It's subtle, but the use of `SameKind` in `(*>@#@$###)` causes `k` to become equal to `Type` (where `k` is the kind of `a` and `b`). This is similar to the reason why defining unsaturated defunctionalization symbols as type families makes this work out (`natMinus` issues notwithstanding). I made a mention of that later in the Note, but it's worth pointing out here as well. I've updated the prose to do just that.